compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -5,46 +5,125 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Common/EncodingUtils.h"
+#include "llvm/Support/Debug.h"
 #include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Utils/IndexingUtils.h"
-
+#include "mlir/IR/BuiltinTypes.h"
 namespace mlir::iree_compiler {
 
 using IREE::LinalgExt::EncodingAttr;
 using IREE::LinalgExt::EncodingRole;
 
+// If tensorType has the encoding of a matmul RESULT with narrow N, returns
+// the transposed type. Otherwise, just returns tensorType.
+static RankedTensorType transposeIfNarrowNResult(RankedTensorType tensorType) {
+  EncodingAttr encoding =
+      tensorType.getEncoding().dyn_cast_or_null<EncodingAttr>();
+  if (!encoding || tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  if (!isNarrowNResult(encoding)) {
+    return tensorType;
+  }
+  if (tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  auto newRole = encoding.getRole().getValue();
+  TypeAttr originalTypeAttr = encoding.getOriginalType();
+  RankedTensorType originalType;
+  if (originalTypeAttr) {
+    originalType = originalTypeAttr.getValue().dyn_cast<RankedTensorType>();
+  }
+  if (!originalType) {
+    originalType = tensorType;
+  }
```

Comment:
We can avoid the second `if` if we initialize `originalType` to `tensorType`. Also, let's use `dyn_cast<T>(x)` here.

```suggestion
  RankedTensorType originalType = tensorType;
  if (originalTypeAttr) {
    originalType = dyn_cast<RankedTensorType>(originalTypeAttr.getValue());
  }
```